### ISSUE TITLE
Attempting to bdr_init_copy a second database fails with Cannot run UPDATE or DELETE on table pg_shseclabel

### ISSUE THREAD

**Issue by jtorral** [2ndQuadrant/bdr#89]
At a very high level view here is what is happening

Node 1
---

```
create database bdrkeeper;
\c bdrkeeper;
create extension btree_gist;
create extension bdr;
bdr_group_create ( local_node_name := 'dbadevpgha03', node_external_dsn := 'host=10.5.101.176 port=5432 dbname=bdrkeeper' );
```

Node 2
---

```
cd $PGDATADIR
rm -rf *
bdr_init_copy -D /var/lib/pgsql/9.4-bdr/data -n dbadevpgha04 -d "host=10.5.101.176 dbname=bdrkeeper port=5432"  --local-dbname="host=10.5.101.179 dbname=bdrkeeper"

bdr_init_copy: starting ...
Getting remote server identification ...
Detected 1 BDR database(s) on remote server
Updating BDR configuration on the remote node:
 bdrkeeper: validating BDR configuration ...
 bdrkeeper: creating replication slot ...
 bdrkeeper: creating node entry for local node ...
Creating base backup of the remote node...
36545/36545 kB (100%), 1/1 tablespace
Creating restore point on remote node ...
Bringing local node to the restore point ...
Transaction log reset
Initializing BDR on the local node:
 bdrkeeper: adding the database to BDR cluster ...
All done
```

So far so good !

Node 1
---

```
create database jorge;
\c jorge;
create extension btree_gist;
create extension bdr;
bdr_group_create ( local_node_name := 'dbadevpgha03', node_external_dsn := 'host=10.5.101.176 port=5432 dbname=jorge' );
```

node 2
---

```
bdr_init_copy: starting ...
Getting remote server identification ...
Detected 2 BDR database(s) on remote server
Updating BDR configuration on the remote node:
 bdrkeeper: validating BDR configuration ...
 bdrkeeper: creating replication slot ...
 bdrkeeper: creating node entry for local node ...
 jorge: validating BDR configuration ...
 jorge: creating replication slot ...
 jorge: creating node entry for local node ...
Creating restore point on remote node ...
Bringing local node to the restore point ...
pg_ctl: another server might be running; trying to start server anyway
Could not update security label: ERROR:  Cannot run UPDATE or DELETE on table pg_shseclabel because it does not have primary key.
HINT:  Add primary key to the table
```

Any help would be appreciated.

Thanks

**Comment by ringerc** [comment]
What you've done here doesn't make much sense. You seem to have:

* Created one DB `bdrkeeper` on node1
* Used `bdr_init_copy` to start a second node based on it
* Gone back to node 1 and created a second BDR-enabled db `jorge`
* .. then used `bdr_init_copy` to copy that to node2. Possibly in the same data directory. It's hard to tell since you didn't show the actual commands run.

This will not work. You just can't do that. `bdr_init_copy` can only be used to bring up a new base backup from the upstream node. Ideally it should detect the case where it's run on an already-established node and refuse to run with an informative error.

If my summary here is correct then what you're doing cannot possibly work, and the only bug here is that we don't prevent you from doing it.

**Comment by ringerc** [comment]
Also seems to have been sent as a mailing list post, see CACut7uS5Ut_pAnpjd9y5AS4y2X+Cc+pVU9_5NY1EeXvMT_YKow@mail.gmail.com

**Comment by jtorral** [comment]
Thanks for the reply.

I am a bit confused with some of your statements.  But, clearly understand
I should run bdr_join_group from sql since there does appear to be an
issues with bdr_init_copy with multiple databases.

What I need clarification on is the following:

Lets say I setup brand new pg cluster. nothing fancy, one table space one
PGDATADIR. Then I run the following:

create database christian;

create database jorge;


\c christian;

create extension btree_gist;

create extension bdr;

select bdr.bdr_group_create ( local_node_name := 'dbadevpgha03',
node_external_dsn := 'host=10.5.101.176 port=5432 dbname=christian' );

SELECT bdr.bdr_node_join_wait_for_ready();


\c jorge;

create extension btree_gist;

create extension bdr;

select bdr.bdr_group_create ( local_node_name := 'dbadevpgha03',
node_external_dsn := 'host=10.5.101.176 port=5432 dbname=jorge' );

SELECT bdr.bdr_node_join_wait_for_ready();


Both databases are in the same $PGDATA cluster. Lets Assume both databases
are production and have many rows in them. How would I get them to the 2nd
node and start bdr? From what you said in one of your messages it sounds
like I cannot do that because they are in the same $PGDATA dir.

If I read your statement wrong, then I could do a pg_basebackup of the
cluster and join each database via sql join group?

In a nut shell, how would I join a brand new downstream node to two active
production databases running on the same server within the same cluster.?


Thanks


JT



On Thu, Jun 18, 2015 at 10:21 PM, Craig Ringer <notifications@github.com>
wrote:

> Also seems to have been sent as a mailing list post, see
> CACut7uS5Ut_pAnpjd9y5AS4y2X+Cc+pVU9_5NY1EeXvMT_YKow@mail.gmail.com
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/2ndQuadrant/bdr/issues/89#issuecomment-113374413>.
>



-- 
Thanks,

Jorge Torralba
----------------------------

Note: This communication may contain privileged or other confidential
information. If you are not the intended recipient, please do not print,
copy, retransmit, disseminate or otherwise use the information. Please
indicate to the sender that you have received this email in error and
delete the copy you received. Thank You.

**Comment by ringerc** [comment]
You seem to be confusing SQL-level logical join using `bdr_group_join` and basebackup-based "physical" join using `bdr_init_copy`.


To add a second (or n'th) node, you either:

* `initdb` a new *blank* cluster
* `CREATE DATABASE` a blank database
* `CREATE EXTENSION btree_gist; CREATE EXTENSION bdr;`
* ... and use `bdr.bdr_group_join` to clone the initial state of an existing bdr node

*or*

* Use `bdr_init_copy` to make a `pg_basebackup` of an existing running node and bring it up as a new BDR node.

One or the other.

If there are multiple independent BDR-enabled databases on the existing node being initialized from, for the SQL-level approach you just join each empty downstream database separately, i.e. create a new database for each, then `bdr.bdr_group_join` each database to the same database on the corresponding existing active bdr node. 

You've observed that there are issues with `bdr_init_copy` and multiple databases. That should be fixed in commit d7fd1cb021e18f1e4747c97c51981382ac8b5e52 . In the mean time you can use the first method, SQL-level join each database individually.

**Comment by ringerc** [comment]
See also #88
